zoo/checklists/views.py

```diff
@@ -0,0 +1,127 @@
+            {
+                "name": service.name,
+                "owner_slug": service.owner_slug,
+                "name_slug": service.name_slug,
+                "completed_checklist_steps": service.checkmarks.count,
+                "total_checklist_steps": sum(
+                    [len(steps) for tag, steps in get_service_steps(service)]
+                ),
+            }
+            for service in self.get_queryset()
+        ]
+        return context
+
+
+class ServiceChecklistView(TemplateView):
+    template_name = "checklists/service_checklist.html"
+
+    def get_context_data(self, **kwargs):
+        service_owner_slug = kwargs.get("service_owner_slug")
+        service_name_slug = kwargs.get("service_name_slug")
+
+        if None in [service_owner_slug, service_name_slug]:
+            raise SuspiciousOperation
+
+        context = super().get_context_data(**kwargs)
+        try:
+            service = service_models.Service.objects.get(
+                owner_slug=service_owner_slug, name_slug=service_name_slug
+            )
+        except service_models.Service.DoesNotExist:
+            raise Http404
+
+        context["service"] = service
+        context["steps"] = [
+            {
+                "tag": tag,
+                "name": next(iter(steps.values())).category_name,
+                "steps": [
+                    attr.evolve(
+                        step,
+                        is_checked=service.checkmarks.filter(step_key=key).exists(),
+                    )
+                    for key, step in steps.items()
+                ],
+            }
+            for tag, steps in get_service_steps(service)
+        ]
+
+        return context
+
+
+class UpdateServiceCheklistForm(forms.Form):
+    checked = forms.BooleanField(required=False)
+
+
+def update_service_checklist(request, checklist_item_key, *args, **kwargs):
+    if request.method != "POST":
+        return JsonResponse({"error": "Method not allowed"}, status=403)
+
+    request_data_form = UpdateServiceCheklistForm(request.POST)
+
+    if not request_data_form.is_valid():
+        return JsonResponse(
+            {"error": request_data_form.errors.as_json(escape_html=True)}, status=400
+        )
+
+    step_checked = request_data_form["checked"].value()
+    owner_slug = kwargs.get("service_owner_slug")
+    name_slug = kwargs.get("service_name_slug")
+
+    if step_checked:
+        try:
+            service = service_models.Service.objects.get(
+                owner_slug=owner_slug, name_slug=name_slug
+            )
+            checkmark = models.Checkmark(service=service, step_key=checklist_item_key)
+            checkmark.full_clean()
+            checkmark.save()
+            return JsonResponse({}, status=200)
+
+        except service_models.Service.DoesNotExist:
+            return JsonResponse(
+                {"error": "The service specified does not exist"}, status=400
+            )
+
+    try:
```

Comment:
I'd indent all the rest under an `else:` even though it doesn't make a difference for the code execution. It will prevent destructive bugs that can stem from extending the previous part with new branches and forgetting to return.

If we end up extending this code more, it also will make it a more sensible switch once we split the two blocks code to `add_mark()` and `remove_mark()` methods.